tests/TestCase/Middleware/ThrottleMiddlewareTest.php

```diff
@@ -0,0 +1,291 @@
+
+        Cache::delete('test-identifier', 'throttle');
+        Cache::drop('throttle');
+    }
+
+    /**
+     * Test if proper string is returned for use as cache expiration key.
+     */
+    public function testGetCacheExpirationKeyMethod()
+    {
+        $middelware = new ThrottleMiddleware();
+        $reflection = $this->getReflection($middelware, '_getCacheExpirationKey', '_identifier');
+
+        // test ip-adress based expiration key (Throttle default)
+        $reflection->property->setValue($middelware, '10.33.10.10');
+        $expected = '10.33.10.10_expires';
+        $result = $reflection->method->invokeArgs($middelware, []);
+        $this->assertEquals($expected, $result);
+
+        // test long JWT Bearer Token based expiration key
+        $reflection->property->setValue($middelware, 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6IjJiMzdhMzVhLTAxNWEtNGUzMi04YTUyLTYzZjQ3ODBkNjY1NCIsImV4cCI6MTQzOTAzMjQ5OH0.U6PkSf6IfSc-o-14UiGy4Rbr9kqqETCKOclf92PXwHY');
+        $expected = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6IjJiMzdhMzVhLTAxNWEtNGUzMi04YTUyLTYzZjQ3ODBkNjY1NCIsImV4cCI6MTQzOTAzMjQ5OH0.U6PkSf6IfSc-o-14UiGy4Rbr9kqqETCKOclf92PXwHY_expires';
+        $result = $reflection->method->invokeArgs($middelware, []);
+        $this->assertEquals($expected, $result);
+    }
+
+    /**
+     * Test x-headers
+     */
+    public function testSetHeadersMethod()
+    {
+        // test disabled headers, should return null
+        $middleware = new ThrottleMiddleware([
+            'headers' => false
+        ]);
+        $reflection = $this->getReflection($middleware, '_setHeaders');
+        $result = $reflection->method->invokeArgs($middleware, [new Response()]);
+        $this->assertInstanceOf('Cake\Http\Response', $result);
+    }
+
+    /**
+     * Test x-headers
+     */
+    public function testRemainingConnectionsMethod()
+    {
+        $middleware = new ThrottleMiddleware();
+        $reflection = $this->getReflection($middleware, '_getRemainingConnections', '_count');
+
+        $reflection->property->setValue($middleware, 7);
+        $result = $reflection->method->invokeArgs($middleware, []);
+        $this->assertEquals('3', $result);
+
+        $reflection->property->setValue($middleware, 0);
+        $result = $reflection->method->invokeArgs($middleware, []);
+        $this->assertEquals('10', $result);
+
+        $reflection->property->setValue($middleware, 10);
+        $result = $reflection->method->invokeArgs($middleware, []);
+        $this->assertEquals('0', $result);
+
+        $reflection->property->setValue($middleware, 11);
+        $result = $reflection->method->invokeArgs($middleware, []);
+        $this->assertEquals('0', $result);
+    }
+
+    /**
+     * Convenience function to return an object with reflection class, accessible
+     * protected method and optional accessible protected property.
+     */
+    public function getReflection($object, $method = false, $property = false)
```

Comment:
That's an historical bad on my behalf whilst I was learning how to test, I dropped this in favor of https://github.com/FriendsOfCake/cakephp-test-utilities